File: 社区相关/LeetCode/CodeStash/278.第一个错误的版本.py
#
# @lc app=leetcode.cn id=278 lang=python3
#
# [278] 第一个错误的版本
#

# @lc code=start
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        left, right = 1, n
        while left <= right:
            pivot = left + (right - left) // 2
            if isBadVersion(pivot):
                if isBadVersion(pivot - 1) == False:
                    return pivot
                else:
                    right = pivot - 1
            if isBadVersion(pivot) == False:
                left = pivot + 1
        
        
# @lc code=end

File: 研究方向/算法/LeetCode/CodeStash/278.第一个错误的版本.cpp
/*
 * @lc app=leetcode.cn id=278 lang=cpp
 *
 * [278] 第一个错误的版本
 */

// @lc code=start
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int left=1, right=n;
        while(left<=right){
            int pivot = left + (right - left) / 2;  // 直接做加法容易溢出
            if(isBadVersion(pivot)) right = pivot - 1;
            else left = pivot + 1;
        }
        return left;
    }
};
// @lc code=end

